transport/http2_client.go

```diff
@@ -554,17 +569,32 @@ func (t *http2Client) handleSettings(f *http2.SettingsFrame) {
 	}
 	f.ForeachSetting(func(s http2.Setting) error {
 		if v, ok := f.Value(s.ID); ok {
-			t.mu.Lock()
-			defer t.mu.Unlock()
 			switch s.ID {
 			case http2.SettingMaxConcurrentStreams:
-				t.maxStreams = v
+				// TODO(zhaoq): This is a hack to avoid significant refactoring of the
+				// code to deal with the unrealistic int32 overflow. Probably will try
+				// to find a better way to handle this later.
+				if v > math.MaxInt32 {
+					v = math.MaxInt32
+				}
+				t.mu.Lock()
+				reset := t.streamsQuota != nil
+				ms := t.maxStreams
+				t.maxStreams = int(v)
+				t.mu.Unlock()
+				if !reset {
+					t.streamsQuota = newQuotaPool(int(v))
+				} else {
+					t.streamsQuota.reset(int(v) - ms)
```

Comment:
On Thu, May 28, 2015 at 4:28 PM, Bryan C. Mills <notifications@github.com>
wrote:

> In transport/http2_client.go
> <https://github.com/grpc/grpc-go/pull/205#discussion_r31287893>:
>
> > -				t.maxStreams = v
> > +				// TODO(zhaoq): This is a hack to avoid significant refactoring of the
> > +				// code to deal with the unrealistic int32 overflow. Probably will try
> > +				// to find a better way to handle this later.
> > +				if v > math.MaxInt32 {
> > +					v = math.MaxInt32
> > +				}
> > +				t.mu.Lock()
> > +				reset := t.streamsQuota != nil
> > +				ms := t.maxStreams
> > +				t.maxStreams = int(v)
> > +				t.mu.Unlock()
> > +				if !reset {
> > +					t.streamsQuota = newQuotaPool(int(v))
> > +				} else {
> > +					t.streamsQuota.reset(int(v) - ms)
>
> This still doesn't work, as far as I can tell. The reset is atomic, but it
> doesn't take into account the number of streams in-flight.
>
Did you read the example I provided before:

*For example, the new quota is 10 (v) and the old quota is 100
(t.maxStream). And line 255 wants to return 99 back to stream quota (100
-1). If the reset occurs before the add, we get t.streamsQuota == -90 after
the reset and 9 quota on the t.streamsQuota.acquire() and t.streamsQuota ==
9, which are all expected. Note that v-t.maxStreams is not the configured
amount. Instead v is -- v - t.maxStreams + x < v because 0 < x <=
t.maxStreams -1.*

If you still think this approach does not count the number of streams
in-flight, could you provide an example too? Note that this does remember
the old value which is t.maxStreams (the new value is v).

> If you want to swap out t.streamsQuota, you need to ensure that there are
> no streams in-flight. Otherwise, you'll need to implement this in terms of
> acquire and/or add.
>
> One alternative to consider: if you store the original quota size, then
> you can compute the delta between the new and old settings. If new > old,
> you can just add it; if new < old, you can acquire the difference and just
> not add it back.
>
> But that makes the handleSettings function take an arbitrarily long time
> while it waits for the existing streams to settle...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/205/files#r31287893>.
>